// course-service/src/courseRepository.js
// Làm mới cho "/services/course-service/src/courseRepository.js":
const TABLE_NAME = "courses";
const PROCESSED_EVENTS_TABLE = "processed_events";

function createError(message, code) {
  const error = new Error(message);
  error.code = code;
  return error;
}

export function createCourseRepository(db) {
  return {
    async findById(id) {
      return db(TABLE_NAME)
        .select(
          "id",
          "title",
          "description",
          "status",
          "enrolled_count",
          "capacity"
        )
        .where({ id })
        .first();
    },

    async findAll({ limit, offset }) {
      return db(TABLE_NAME)
        .select(
          "id",
          "title",
          "description",
          "status",
          "enrolled_count",
          "capacity"
        )
        .orderBy("id", "asc")
        .limit(limit)
        .offset(offset);
    },

    async countAll() {
      const row = await db(TABLE_NAME).count({ count: "*" }).first();
      return Number(row.count);
    },

    async applyEnrollmentConfirmed(event) {
  const { eventId, eventType, correlationId, payload } = event;

  return db.transaction(async (trx) => {
    const insertedEvents = await trx(PROCESSED_EVENTS_TABLE)
      .insert({
        event_id: eventId,
        event_type: eventType,
        processed_at: trx.fn.now()
      })
      .onConflict("event_id")
      .ignore()
      .returning("event_id");

    if (insertedEvents.length === 0) {
      return {
        duplicated: true,
        message: `Event ${eventId} was already processed`
      };
    }

    const updatedCourses = await trx(TABLE_NAME)
      .where({ id: payload.courseId })
      .increment("enrolled_count", 1)
      .returning([
        "id",
        "title",
        "description",
        "enrolled_count"
      ]);

    if (updatedCourses.length === 0) {
      throw createError(`COURSE_NOT_FOUND: ${payload.courseId}`, "NOT_FOUND");
    }

    const updatedCourse = updatedCourses[0];
    const courseEventId = randomUUID();

    await trx(OUTBOX_EVENTS_TABLE).insert({
      id: courseEventId,
      event_type: "CourseEnrollmentCountIncreased",
      routing_key: "course.enrolled_count.increased",
      version: 1,
      correlation_id: correlationId ?? null,
      payload: {
        courseId: updatedCourse.id,
        courseTitle: updatedCourse.title,
        enrolledCount: updatedCourse.enrolled_count,
        incrementBy: 1,
        enrollmentId: payload.enrollmentId,
        studentId: payload.studentId
      },
      status: "pending",
      attempts: 0
    });

    return {
      duplicated: false,
      course: updatedCourse,
      emittedEventId: courseEventId
    };
  });
}
  };
}
